packages/dds/tree/src/feature-libraries/indexing/anchorTreeIndex.ts

```diff
@@ -0,0 +1,417 @@
+ * A key can map to multiple nodes but each collection of nodes only results in a single value.
+ *
+ * @remarks
+ * Detached nodes are stored in the index but filtered out when any public facing apis are called. This means that
+ * calling {@link keys} will not include any keys that are stored in the index but only map to detached nodes.
+ */
+export class AnchorTreeIndex<TKey extends TreeIndexKey, TValue>
+	implements TreeIndex<TKey, TValue>
+{
+	/**
+	 * Caches {@link KeyFinder}s for each schema definition. If a schema maps to null, it does not
+	 * need to be considered at all for this index. This allows us to skip subtrees that aren't relevant
+	 * as a performance optimization.
+	 */
+	private readonly keyFinders = new Map<TreeNodeSchemaIdentifier, KeyFinder<TKey> | null>();
+	/**
+	 * The actual index from keys to anchor nodes.
+	 */
+	private readonly nodes = new Map<TKey, AnchorNode[]>();
+	/**
+	 * Maintained for efficient removal of anchor nodes from the index when updating their keys
+	 */
+	private readonly anchorKeys = new Map<AnchorNode, TKey>();
+	/**
+	 * Keeps track of anchors for disposal.
+	 */
+	private readonly anchors = new Map<AnchorNode, Anchor[]>();
+
+	/**
+	 * @param forest - the forest that is being indexed
+	 * @param indexer - a function that retrieves the key finder based on a given schema or undefined if the schema does not have an associated key finder
+	 * @param getValue - a function that returns the value or undefined given at least one anchor node
+	 * @param checkTreeStatus - a function that gets the tree status from an anchor node, used for filtering out detached nodes
+	 */
+	public constructor(
+		private readonly forest: IForestSubscription,
+		private readonly indexer: (
+			schemaId: TreeNodeSchemaIdentifier,
+		) => KeyFinder<TKey> | undefined,
+		private readonly getValue: (anchorNodes: TreeIndexNodes<AnchorNode>) => TValue | undefined,
+		private readonly checkTreeStatus: (node: AnchorNode) => TreeStatus | undefined,
+	) {
+		this.forest.registerAnnouncedVisitor(this.acquireVisitor.bind(this));
+
+		const detachedFieldKeys: FieldKey[] = [];
+		const detachedFieldsCursor = forest.getCursorAboveDetachedFields();
+		forEachField(detachedFieldsCursor, (field) => {
+			detachedFieldKeys.push(field.getFieldKey());
+		});
+
+		// index all existing trees (this includes the primary document tree and all other detached/removed trees)
+		for (const fieldKey of detachedFieldKeys) {
+			const cursor = forest.allocateCursor();
+			forest.tryMoveCursorToField({ fieldKey, parent: undefined }, cursor);
+			this.indexField(cursor);
+			cursor.free();
+		}
+	}
+
+	/**
+	 * Creates an announced visitor that responds to edits to the forest and updates the index accordingly.
+	 */
+	private acquireVisitor(): AnnouncedVisitor {
+		let parentField: FieldKey | undefined;
+		let parent: UpPath | undefined;
+
+		return createAnnouncedVisitor({
```

Comment:
A top level description here talking about why each method in the visitor does what it does would be really useful for maintainability IMO.